File: backend/app/routers/graines.py
"""Routers pour Graine et PackGraine"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from app.database import get_db
from app.models import Graine, PackGraine, Breeder, Variete, Croisement
from app.schemas.graine import (
    GraineCreate,
    GraineRead,
    GraineSimple,
    PackGraineCreate,
    PackGraineRead,
    PackGraineCompletCreate,
    PackGraineCompletRead,
    PackGraineCompletUpdate,
    CatalogueItem,
)

router = APIRouter(prefix="/api", tags=["graines"])


# ========== Pack Graine ==========

@router.get("/packs", response_model=list[PackGraineRead])
def get_packs(db: Session = Depends(get_db)):
    return db.query(PackGraine).all()


@router.get("/packs/{pack_id}", response_model=PackGraineRead)
def get_pack(pack_id: int, db: Session = Depends(get_db)):
    pack = db.query(PackGraine).filter(PackGraine.id_packgraine == pack_id).first()
    if not pack:
        raise HTTPException(status_code=404, detail="Pack non trouvé")
    return pack


@router.get("/packs/{pack_id}/graines", response_model=list[GraineSimple])
def get_graines_du_pack(pack_id: int, db: Session = Depends(get_db)):
    pack = db.query(PackGraine).filter(PackGraine.id_packgraine == pack_id).first()
    if not pack:
        raise HTTPException(status_code=404, detail="Pack non trouvé")
    return db.query(Graine).filter(Graine.id_packgraine == pack_id).order_by(Graine.id_graine).all()


@router.post("/packs", response_model=PackGraineRead)
def create_pack(pack: PackGraineCreate, db: Session = Depends(get_db)):
    db_pack = PackGraine(
        id_fournisseur=pack.id_fournisseur,
        nbr_graines=pack.nbr_graines,
        prix_achat=pack.prix_achat,
        date_achat=pack.date_achat,
    )
    db.add(db_pack)
    db.commit()
    db.refresh(db_pack)
    return db_pack


@router.post("/packs/complet", response_model=PackGraineCompletRead)
def create_pack_complet(pack: PackGraineCompletCreate, db: Session = Depends(get_db)):
    """Crée un pack + toutes ses graines. Met à jour le croisement de la variété si fourni."""
    breeder = db.query(Breeder).filter(Breeder.id_breeder == pack.id_breeder).first()
    if not breeder:
        raise HTTPException(status_code=404, detail="Breeder non trouvé")

    variete = db.query(Variete).filter(Variete.id_variete == pack.id_variete).first()
    if not variete:
        raise HTTPException(status_code=404, detail="Variété non trouvée")

    # Met à jour le croisement si fourni
    if pack.croisement_variete is not None:
        variete.croisement_variete = pack.croisement_variete

    db_pack = PackGraine(
        id_fournisseur=pack.id_fournisseur,
        nbr_graines=pack.nbr_graines,
        prix_achat=pack.prix_achat,
        date_achat=pack.date_achat,
    )
    db.add(db_pack)
    db.flush()

    prix_par_graine = None
    if pack.prix_achat and pack.nbr_graines > 0:
        prix_par_graine = float(pack.prix_achat) / pack.nbr_graines

    for _ in range(pack.nbr_graines):
        db_graine = Graine(
            id_breeder=pack.id_breeder,
            id_variete=pack.id_variete,
            id_packgraine=db_pack.id_packgraine,
            types_graines=pack.types_graines,
            duree_flo_min=pack.duree_flo_min,
            duree_flo_max=pack.duree_flo_max,
            prix_achat=prix_par_graine,
            edition_limite=pack.edition_limite,
            date_achat=pack.date_achat,
            utilisee=False,
        )
        db.add(db_graine)

    db.commit()
    db.refresh(db_pack)

    return PackGraineCompletRead(
        id_packgraine=db_pack.id_packgraine,
        nbr_graines=pack.nbr_graines,
        nbr_graines_crees=pack.nbr_graines,
        breeder_nom=breeder.nom_breeder,
        variete_nom=variete.nom_variete,
    )


@router.put("/packs/{pack_id}/complet", response_model=PackGraineCompletRead)
def update_pack_complet(pack_id: int, pack: PackGraineCompletUpdate, db: Session = Depends(get_db)):
    """Met à jour pack + graines + croisement de la variété."""
    db_pack = db.query(PackGraine).filter(PackGraine.id_packgraine == pack_id).first()
    if not db_pack:
        raise HTTPException(status_code=404, detail="Pack non trouvé")

    breeder = db.query(Breeder).filter(Breeder.id_breeder == pack.id_breeder).first()
    if not breeder:
        raise HTTPException(status_code=404, detail="Breeder non trouvé")

    variete = db.query(Variete).filter(Variete.id_variete == pack.id_variete).first()
    if not variete:
        raise HTTPException(status_code=404, detail="Variété non trouvée")

    # Met à jour le croisement de la variété si fourni
    if pack.croisement_variete is not None:
        variete.croisement_variete = pack.croisement_variete

    # Met à jour le pack
    db_pack.id_fournisseur = pack.id_fournisseur
    db_pack.prix_achat = pack.prix_achat
    db_pack.date_achat = pack.date_achat

    # Met à jour les métadonnées de toutes les graines existantes
    graines = db.query(Graine).filter(Graine.id_packgraine == pack_id).all()
    for g in graines:
        g.id_breeder = pack.id_breeder
        g.id_variete = pack.id_variete
        g.types_graines = pack.types_graines
        g.duree_flo_min = pack.duree_flo_min
        g.duree_flo_max = pack.duree_flo_max
        g.edition_limite = pack.edition_limite

    # Ajustement du nombre de graines si demandé
    if pack.nbr_graines is not None and pack.nbr_graines > 0:
        current_total = len(graines)
        new_total = pack.nbr_graines

        if new_total > current_total:
            # Ajouter des graines manquantes
            prix_par_graine = None
            if pack.prix_achat and new_total > 0:
                prix_par_graine = float(pack.prix_achat) / new_total
            for _ in range(new_total - current_total):
                db.add(Graine(
                    id_breeder=pack.id_breeder,
                    id_variete=pack.id_variete,
                    id_packgraine=pack_id,
                    types_graines=pack.types_graines,
                    duree_flo_min=pack.duree_flo_min,
                    duree_flo_max=pack.duree_flo_max,
                    prix_achat=prix_par_graine,
                    edition_limite=pack.edition_limite,
                    date_achat=pack.date_achat,
                    utilisee=False,
                ))

        elif new_total < current_total:
            # Supprimer des graines — non-utilisées en premier
            to_remove = current_total - new_total
            non_utilisees = (
                db.query(Graine)
                .filter(Graine.id_packgraine == pack_id, Graine.utilisee == False)
                .order_by(Graine.id_graine.desc())
                .limit(to_remove)
                .all()
            )
            for g in non_utilisees:
                db.delete(g)
                to_remove -= 1
            # Si toujours trop, supprimer des utilisées
            if to_remove > 0:
                utilisees = (
                    db.query(Graine)
                    .filter(Graine.id_packgraine == pack_id, Graine.utilisee == True)
                    .order_by(Graine.id_graine.desc())
                    .limit(to_remove)
                    .all()
                )
                for g in utilisees:
                    db.delete(g)

        db_pack.nbr_graines = new_total
        db.flush()

    final_count = db.query(Graine).filter(Graine.id_packgraine == pack_id).count()

    db.commit()
    db.refresh(db_pack)

    return PackGraineCompletRead(
        id_packgraine=db_pack.id_packgraine,
        nbr_graines=db_pack.nbr_graines,
        nbr_graines_crees=final_count,
        breeder_nom=breeder.nom_breeder,
        variete_nom=variete.nom_variete,
    )


@router.delete("/packs/{pack_id}")
def delete_pack(pack_id: int, db: Session = Depends(get_db)):
    db_pack = db.query(PackGraine).filter(PackGraine.id_packgraine == pack_id).first()
    if not db_pack:
        raise HTTPException(status_code=404, detail="Pack non trouvé")
    # Effacer la référence dans les croisements avant suppression (FK constraint)
    db.query(Croisement).filter(
        Croisement.id_packgraine_resultat == pack_id
    ).update({"id_packgraine_resultat": None}, synchronize_session=False)
    db.query(Graine).filter(Graine.id_packgraine == pack_id).delete(synchronize_session=False)
    db.delete(db_pack)
    db.commit()
    return {"message": "Pack supprimé"}


# ========== Graine individuelle ==========

@router.get("/graines", response_model=list[GraineRead])
def get_graines(db: Session = Depends(get_db)):
    return db.query(Graine).all()


@router.get("/graines/{graine_id}", response_model=GraineRead)
def get_graine(graine_id: int, db: Session = Depends(get_db)):
    graine = db.query(Graine).filter(Graine.id_graine == graine_id).first()
    if not graine:
        raise HTTPException(status_code=404, detail="Graine non trouvée")
    return graine


@router.post("/graines", response_model=GraineRead)
def create_graine(graine: GraineCreate, db: Session = Depends(get_db)):
    db_graine = Graine(**graine.model_dump())
    db.add(db_graine)
    db.commit()
    db.refresh(db_graine)
    return db_graine


@router.put("/graines/{graine_id}", response_model=GraineRead)
def update_graine(graine_id: int, graine: GraineCreate, db: Session = Depends(get_db)):
    db_graine = db.query(Graine).filter(Graine.id_graine == graine_id).first()
    if not db_graine:
        raise HTTPException(status_code=404, detail="Graine non trouvée")
    for k, v in graine.model_dump().items():
        setattr(db_graine, k, v)
    db.commit()
    db.refresh(db_graine)
    return db_graine


@router.patch("/graines/{graine_id}/toggle", response_model=GraineSimple)
def toggle_graine_utilisee(graine_id: int, db: Session = Depends(get_db)):
    db_graine = db.query(Graine).filter(Graine.id_graine == graine_id).first()
    if not db_graine:
        raise HTTPException(status_code=404, detail="Graine non trouvée")
    db_graine.utilisee = not db_graine.utilisee
    db.commit()
    db.refresh(db_graine)
    return db_graine


@router.delete("/graines/{graine_id}")
def delete_graine(graine_id: int, db: Session = Depends(get_db)):
    db_graine = db.query(Graine).filter(Graine.id_graine == graine_id).first()
    if not db_graine:
        raise HTTPException(status_code=404, detail="Graine non trouvée")
    db.delete(db_graine)
    db.commit()
    return {"message": "Graine supprimée"}


# ========== Catalogue ==========

@router.get("/catalogue", response_model=list[CatalogueItem])
def get_catalogue(db: Session = Depends(get_db)):
    """Catalogue complet. paquet_ouvert = au moins une graine utilisée."""
    packs = db.query(PackGraine).all()
    catalogue = []

    for pack in packs:
        # Compter les graines NON utilisées (utilisee = False ou NULL)
        remaining = db.query(func.count(Graine.id_graine)).filter(
            Graine.id_packgraine == pack.id_packgraine,
            (Graine.utilisee == False) | (Graine.utilisee == None)
        ).scalar() or 0

        first_graine = db.query(Graine).filter(
            Graine.id_packgraine == pack.id_packgraine
        ).first()

        if first_graine:
            breeder = db.query(Breeder).filter(Breeder.id_breeder == first_graine.id_breeder).first() if first_graine.id_breeder else None
            variete = db.query(Variete).filter(Variete.id_variete == first_graine.id_variete).first() if first_graine.id_variete else None
            if pack.id_packgraine >= 130:  # debug packs récents
                print(f"[DEBUG CATALOGUE] pack={pack.id_packgraine} nbr_graines={pack.nbr_graines} remaining={remaining} graine.id_variete={first_graine.id_variete} variete_nom={variete.nom_variete if variete else 'None'} breeder_nom={breeder.nom_breeder if breeder else 'None'}", flush=True)
            catalogue.append(CatalogueItem(
                id_packgraine=pack.id_packgraine,
                id_fournisseur=pack.id_fournisseur,
                id_breeder=first_graine.id_breeder,
                id_variete=first_graine.id_variete,
                breeder_nom=breeder.nom_breeder if breeder else "Inconnu",
                variete_nom=variete.nom_variete if variete else "Inconnue",
                croisement_variete=variete.croisement_variete if variete else None,
                lien_web=variete.lien_web if variete else None,
                type_graines=first_graine.types_graines,
                duree_flo_min=first_graine.duree_flo_min,
                duree_flo_max=first_graine.duree_flo_max,
                prix_par_graine=float(pack.prix_achat / pack.nbr_graines) if pack.prix_achat and pack.nbr_graines else None,
                nbr_graines_total=pack.nbr_graines or 0,
                nbr_graines_restantes=remaining,
                paquet_ouvert=(remaining < (pack.nbr_graines or 0)),
                edition_limite=first_graine.edition_limite,
                date_achat=pack.date_achat,
            ))

    return catalogue

File: frontend/src/local/handlers/graines.ts
// ─── Handlers locaux : Graine, PackGraine, Catalogue ──────────────────────────
// Miroir de backend/app/routers/graines.py
import { route } from '../router'
import { query, one, oneOr404, count, insert, updateById, run, boolify } from '../helpers'

type Row = Record<string, unknown>

const GRAINE_BOOLS = ['edition_limite', 'utilisee']

async function graineEnrichie(row: Row): Promise<Row> {
  boolify(row, GRAINE_BOOLS)
  const breeder = row.id_breeder
    ? await one('SELECT * FROM "Breeder" WHERE id_breeder = ?', [row.id_breeder]) : null
  const variete = row.id_variete
    ? await one('SELECT * FROM "Variete" WHERE id_variete = ?', [row.id_variete]) : null
  return { ...row, breeder, variete }
}

// ── Packs ─────────────────────────────────────────────────────────────────────
// NB : '/packs/complet' enregistré avant '/packs/:id'.

route('GET', '/packs', async () => ({ data: await query('SELECT * FROM "PackGraine"') }))

route('POST', '/packs/complet', async ({ body }) => {
  const p = body as Row
  const breeder = await oneOr404<Row>('SELECT * FROM "Breeder" WHERE id_breeder = ?',
    [p.id_breeder], 'Breeder non trouvé')
  const variete = await oneOr404<Row>('SELECT * FROM "Variete" WHERE id_variete = ?',
    [p.id_variete], 'Variété non trouvée')

  if (p.croisement_variete !== null && p.croisement_variete !== undefined) {
    await run('UPDATE "Variete" SET croisement_variete = ? WHERE id_variete = ?',
      [p.croisement_variete, p.id_variete])
  }

  const nbr = Number(p.nbr_graines)
  const idPack = await insert('PackGraine', {
    id_fournisseur: p.id_fournisseur ?? null,
    nbr_graines: nbr,
    prix_achat: p.prix_achat ?? null,
    date_achat: p.date_achat ?? null,
  })

  const prixParGraine = p.prix_achat && nbr > 0 ? Number(p.prix_achat) / nbr : null
  for (let i = 0; i < nbr; i++) {
    await insert('Graine', {
      id_breeder: p.id_breeder,
      id_variete: p.id_variete,
      id_packgraine: idPack,
      types_graines: p.types_graines ?? null,
      duree_flo_min: p.duree_flo_min ?? null,
      duree_flo_max: p.duree_flo_max ?? null,
      prix_achat: prixParGraine,
      edition_limite: !!p.edition_limite,
      date_achat: p.date_achat ?? null,
      utilisee: false,
    })
  }

  return {
    data: {
      id_packgraine: idPack,
      nbr_graines: nbr,
      nbr_graines_crees: nbr,
      breeder_nom: breeder.nom_breeder,
      variete_nom: variete.nom_variete,
    },
  }
})

route('GET', '/packs/:id', async ({ params }) => ({
  data: await oneOr404('SELECT * FROM "PackGraine" WHERE id_packgraine = ?', [params.id], 'Pack non trouvé'),
}))

route('GET', '/packs/:id/graines', async ({ params }) => {
  await oneOr404('SELECT * FROM "PackGraine" WHERE id_packgraine = ?', [params.id], 'Pack non trouvé')
  const rows = await query<Row>(
    'SELECT id_graine, utilisee FROM "Graine" WHERE id_packgraine = ? ORDER BY id_graine', [params.id])
  return { data: rows.map(r => boolify(r, ['utilisee'])) }
})

route('POST', '/packs', async ({ body }) => {
  const p = body as Row
  const id = await insert('PackGraine', {
    id_fournisseur: p.id_fournisseur ?? null,
    nbr_graines: p.nbr_graines,
    prix_achat: p.prix_achat ?? null,
    date_achat: p.date_achat ?? null,
  })
  return { data: await one('SELECT * FROM "PackGraine" WHERE id_packgraine = ?', [id]) }
})

route('PUT', '/packs/:id/complet', async ({ params, body }) => {
  const packId = Number(params.id)
  const p = body as Row
  await oneOr404('SELECT * FROM "PackGraine" WHERE id_packgraine = ?', [packId], 'Pack non trouvé')
  const breeder = await oneOr404<Row>('SELECT * FROM "Breeder" WHERE id_breeder = ?',
    [p.id_breeder], 'Breeder non trouvé')
  const variete = await oneOr404<Row>('SELECT * FROM "Variete" WHERE id_variete = ?',
    [p.id_variete], 'Variété non trouvée')

  if (p.croisement_variete !== null && p.croisement_variete !== undefined) {
    await run('UPDATE "Variete" SET croisement_variete = ? WHERE id_variete = ?',
      [p.croisement_variete, p.id_variete])
  }

  await updateById('PackGraine', 'id_packgraine', packId, {
    id_fournisseur: p.id_fournisseur ?? null,
    prix_achat: p.prix_achat ?? null,
    date_achat: p.date_achat ?? null,
  })

  // Màj métadonnées de toutes les graines du pack
  await run(
    `UPDATE "Graine" SET id_breeder = ?, id_variete = ?, types_graines = ?,
     duree_flo_min = ?, duree_flo_max = ?, edition_limite = ?
     WHERE id_packgraine = ?`,
    [p.id_breeder, p.id_variete, p.types_graines ?? null,
     p.duree_flo_min ?? null, p.duree_flo_max ?? null, p.edition_limite ? 1 : 0, packId])

  // Ajustement du nombre de graines
  if (p.nbr_graines !== null && p.nbr_graines !== undefined && Number(p.nbr_graines) > 0) {
    const newTotal = Number(p.nbr_graines)
    const currentTotal = await count('SELECT COUNT(*) AS n FROM "Graine" WHERE id_packgraine = ?', [packId])

    if (newTotal > currentTotal) {
      const prixParGraine = p.prix_achat && newTotal > 0 ? Number(p.prix_achat) / newTotal : null
      for (let i = 0; i < newTotal - currentTotal; i++) {
        await insert('Graine', {
          id_breeder: p.id_breeder,
          id_variete: p.id_variete,
          id_packgraine: packId,
          types_graines: p.types_graines ?? null,
          duree_flo_min: p.duree_flo_min ?? null,
          duree_flo_max: p.duree_flo_max ?? null,
          prix_achat: prixParGraine,
          edition_limite: !!p.edition_limite,
          date_achat: p.date_achat ?? null,
          utilisee: false,
        })
      }
    } else if (newTotal < currentTotal) {
      // Supprimer des graines — non-utilisées en premier, puis utilisées si besoin
      let toRemove = currentTotal - newTotal
      const nonUtilisees = await query<Row>(
        `SELECT id_graine FROM "Graine" WHERE id_packgraine = ? AND (utilisee = 0 OR utilisee IS NULL)
         ORDER BY id_graine DESC LIMIT ?`, [packId, toRemove])
      for (const g of nonUtilisees) {
        await run('DELETE FROM "Graine" WHERE id_graine = ?', [g.id_graine])
        toRemove--
      }
      if (toRemove > 0) {
        const utilisees = await query<Row>(
          `SELECT id_graine FROM "Graine" WHERE id_packgraine = ? AND utilisee = 1
           ORDER BY id_graine DESC LIMIT ?`, [packId, toRemove])
        for (const g of utilisees) {
          await run('DELETE FROM "Graine" WHERE id_graine = ?', [g.id_graine])
        }
      }
    }
    await run('UPDATE "PackGraine" SET nbr_graines = ? WHERE id_packgraine = ?', [newTotal, packId])
  }

  const pack = await one<Row>('SELECT * FROM "PackGraine" WHERE id_packgraine = ?', [packId])
  const finalCount = await count('SELECT COUNT(*) AS n FROM "Graine" WHERE id_packgraine = ?', [packId])
  return {
    data: {
      id_packgraine: packId,
      nbr_graines: pack?.nbr_graines,
      nbr_graines_crees: finalCount,
      breeder_nom: breeder.nom_breeder,
      variete_nom: variete.nom_variete,
    },
  }
})

route('DELETE', '/packs/:id', async ({ params }) => {
  await oneOr404('SELECT * FROM "PackGraine" WHERE id_packgraine = ?', [params.id], 'Pack non trouvé')
  await run('UPDATE "Croisement" SET id_packgraine_resultat = NULL WHERE id_packgraine_resultat = ?', [params.id])
  await run('DELETE FROM "Graine" WHERE id_packgraine = ?', [params.id])
  await run('DELETE FROM "PackGraine" WHERE id_packgraine = ?', [params.id])
  return { data: { message: 'Pack supprimé' } }
})

// ── Graines individuelles ─────────────────────────────────────────────────────

route('GET', '/graines', async () => {
  const rows = await query<Row>('SELECT * FROM "Graine"')
  return { data: await Promise.all(rows.map(graineEnrichie)) }
})

route('GET', '/graines/:id', async ({ params }) => ({
  data: await graineEnrichie(
    await oneOr404('SELECT * FROM "Graine" WHERE id_graine = ?', [params.id], 'Graine non trouvée')),
}))

route('POST', '/graines', async ({ body }) => {
  const g = body as Row
  const id = await insert('Graine', {
    id_breeder: g.id_breeder,
    id_variete: g.id_variete,
    id_packgraine: g.id_packgraine,
    duree_flo_min: g.duree_flo_min ?? null,
    duree_flo_max: g.duree_flo_max ?? null,
    types_graines: g.types_graines ?? null,
    prix_achat: g.prix_achat ?? null,
    edition_limite: !!g.edition_limite,
    date_achat: g.date_achat ?? null,
    utilisee: !!g.utilisee,
  })
  const row = await one<Row>('SELECT * FROM "Graine" WHERE id_graine = ?', [id])
  return { data: await graineEnrichie(row!) }
})

route('PUT', '/graines/:id', async ({ params, body }) => {
  await oneOr404('SELECT * FROM "Graine" WHERE id_graine = ?', [params.id], 'Graine non trouvée')
  const g = body as Row
  await updateById('Graine', 'id_graine', params.id, {
    id_breeder: g.id_breeder,
    id_variete: g.id_variete,
    id_packgraine: g.id_packgraine,
    duree_flo_min: g.duree_flo_min ?? null,
    duree_flo_max: g.duree_flo_max ?? null,
    types_graines: g.types_graines ?? null,
    prix_achat: g.prix_achat ?? null,
    edition_limite: !!g.edition_limite,
    date_achat: g.date_achat ?? null,
    utilisee: !!g.utilisee,
  })
  const row = await one<Row>('SELECT * FROM "Graine" WHERE id_graine = ?', [params.id])
  return { data: await graineEnrichie(row!) }
})

route('PATCH', '/graines/:id/toggle', async ({ params }) => {
  const g = await oneOr404<Row>('SELECT * FROM "Graine" WHERE id_graine = ?', [params.id], 'Graine non trouvée')
  const nouvelle = Number(g.utilisee) ? 0 : 1
  await run('UPDATE "Graine" SET utilisee = ? WHERE id_graine = ?', [nouvelle, params.id])
  return { data: { id_graine: Number(params.id), utilisee: !!nouvelle } }
})

route('DELETE', '/graines/:id', async ({ params }) => {
  await oneOr404('SELECT * FROM "Graine" WHERE id_graine = ?', [params.id], 'Graine non trouvée')
  await run('DELETE FROM "Graine" WHERE id_graine = ?', [params.id])
  return { data: { message: 'Graine supprimée' } }
})

// ── Catalogue ─────────────────────────────────────────────────────────────────

route('GET', '/catalogue', async () => {
  const packs = await query<Row>('SELECT * FROM "PackGraine"')
  const catalogue: Row[] = []

  for (const pack of packs) {
    const remaining = await count(
      'SELECT COUNT(*) AS n FROM "Graine" WHERE id_packgraine = ? AND (utilisee = 0 OR utilisee IS NULL)',
      [pack.id_packgraine])
    const first = await one<Row>(
      'SELECT * FROM "Graine" WHERE id_packgraine = ? ORDER BY id_graine LIMIT 1',
      [pack.id_packgraine])
    if (!first) continue

    const breeder = first.id_breeder
      ? await one<Row>('SELECT * FROM "Breeder" WHERE id_breeder = ?', [first.id_breeder]) : null
    const variete = first.id_variete
      ? await one<Row>('SELECT * FROM "Variete" WHERE id_variete = ?', [first.id_variete]) : null
    const total = Number(pack.nbr_graines ?? 0)

    catalogue.push({
      id_packgraine: pack.id_packgraine,
      id_fournisseur: pack.id_fournisseur ?? null,
      id_breeder: first.id_breeder ?? null,
      id_variete: first.id_variete ?? null,
      breeder_nom: breeder?.nom_breeder ?? 'Inconnu',
      variete_nom: variete?.nom_variete ?? 'Inconnue',
      croisement_variete: variete?.croisement_variete ?? null,
      lien_web: variete?.lien_web ?? null,
      type_graines: first.types_graines ?? null,
      duree_flo_min: first.duree_flo_min ?? null,
      duree_flo_max: first.duree_flo_max ?? null,
      prix_par_graine: pack.prix_achat && total ? Number(pack.prix_achat) / total : null,
      nbr_graines_total: total,
      nbr_graines_restantes: remaining,
      paquet_ouvert: remaining < total,
      edition_limite: !!Number(first.edition_limite ?? 0),
      date_achat: pack.date_achat ?? null,
    })
  }
  return { data: catalogue }
})

